act #20332 Functionality to automatically load library extensions

johanandren opened the pull request:
Refs #20332

johanandren committed:
act #20332 Functionality to automatically load library extensions

ktoso commented on akka-actor-tests/src/test/resources/reference.conf:
```diff
@@ -1,4 +1,7 @@
 akka {
+  # for the ConfigSpec
+  library-extensions += "akka.config.DummyExtension"
```
Hm, ok since we really want to test that exact logic..

johanandren replied:
Couldn't think of a less intrusive way to do :(

ktoso commented on akka-actor-tests/src/test/scala/akka/config/ConfigSpec.scala:
```diff
@@ -151,5 +152,13 @@ class ConfigSpec extends AkkaSpec(ConfigFactory.defaultReference(ActorSystem.fin
+      val listedExtensions = system.settings.config.getStringList("akka.library-extensions")
+      listedExtensions.size should be > 0
+      // could be initalized by other tests, so at least once
+      DummyExtension.createCount.get() should be > 0
```
OK

ktoso commented on akka-actor-tests/src/test/scala/akka/config/DummyExtension.scala:
```diff
@@ -0,0 +1,16 @@
+package akka.config
```
missing license

ktoso commented on akka-actor/src/main/resources/reference.conf:
```diff
@@ -59,6 +59,16 @@ akka {
   log-dead-letters-during-shutdown = on
 
+  # Library extensions are regular extensions that are loaded at startup and are
+  # available for third party libraries to auto-load themselves through their
```
for library authors to enable auto-loading (if on classpath) of extensions?

johanandren replied:
Additionally followed by an unfinished sentence. D'oh.

patriknw commented on akka-actor/src/main/resources/reference.conf:
```diff
@@ -59,6 +59,15 @@ akka {
+  # present on the classpath. This is done by appending entries:
+  # 'library-extensions += "Extension"' in the library `reference.conf`.
+  #
+  # Should not be set by end user applications in 'application.conf'.
```
add ", use the `extensions` property for that"

patriknw replied:
also mention something similar to "List FQCN of extensions which shall be loaded at actor system startup."

patriknw commented on akka-actor/src/main/scala/akka/actor/ActorSystem.scala:
```diff
@@ -779,14 +779,19 @@ private[akka] class ActorSystemImpl(
+          case Success(p: ExtensionIdProvider) ⇒ registerExtension(p.lookup())
+          case Success(p: ExtensionId[_])      ⇒ registerExtension(p)
+          case Success(other)                  ⇒ log.error("[{}] is not an 'ExtensionIdProvider' or 'ExtensionId', skipping...", fqcn)
+          case Failure(problem)                ⇒ log.error(problem, "While trying to load extension [{}], skipping...", fqcn)
```
so, does this mean that it's not possible to fail the startup if the extension throws when being created?

johanandren replied:
That is if loading the `ExtensionId` or `ExtensionIdProvider` fails, not creating the actual `Extension` as far as I understand, will double check though (I think there is a test in config-checker covering this, but I'll add one in the `ConfigSpec` as well).

johanandren replied:
Verified and covered with test, but, maybe loading it should also fail? I mean, you don't list extensions because you want them to maybe be loaded on startup...

patriknw replied:
then we can only change that for `library-extensions`, that is ok with me

johanandren commented:
Realized there were existing extensions loading tests in `akka.actor.ActorSystemSpec` so I extracted those, and combined with my new tests for `library-extensions` and added some additional tests covering error handling of extension loading.

johanandren commented:
I have now manually verified that classpath ordering is not an issue (yay!)

johanandren commented on akka-docs/rst/scala/extending-akka.rst:
```diff
@@ -87,3 +87,21 @@ Use it:
+where the user would ever want it disabled or have specific support for disabling such sub-features. One example where
+this could be important is in tests.
+
+.. warning::
```
This gets a warning from sphinx, so must fix before I can merge.

johanandren commented:
All the things fixed.

patriknw commented:
LGTM

drewhk commented on akka-actor-tests/src/test/scala/akka/actor/ExtensionSpec.scala:
```diff
@@ -0,0 +1,140 @@
+      }
+    }
+
+    "log an error if an extension listed in akka.extensions cannot be loaded" in {
```
why log an error instead of failing the system? It feels weird that if we find one and it fails to start, then we bail out, but if we can't find it we just chug along.

johanandren replied:
I agree, but @patriknw said we can't change it.

patriknw replied:
I agree, but that's current behavior and I think that would be a bit dangerous to change in a patch release. For the the new library-extensions we fail.

drewhk commented:
LGTM
